QEfficient/utils/logging_utils.py

```diff
@@ -6,53 +6,72 @@
 # -----------------------------------------------------------------------------
 
 import logging
+import os
+from datetime import datetime
+from typing import Optional
 
+from concurrent_log_handler import ConcurrentRotatingFileHandler
 
-class QEffFormatter(logging.Formatter):
-    """
-    Formatter class used to set colors for printing different logging levels of messages on console.
-    """
-
-    cyan: str = "\x1b[38;5;14m"
-    yellow: str = "\x1b[33;20m"
-    red: str = "\x1b[31;20m"
-    bold_red: str = "\x1b[31;1m"
-    reset: str = "\x1b[0m"
-    common_format: str = "%(levelname)s - %(name)s - %(message)s"  # type: ignore
-    format_with_line_info = "%(levelname)s - %(name)s - %(message)s  (%(filename)s:%(lineno)d)"  # type: ignore
-
-    FORMATS = {
-        logging.DEBUG: cyan + format_with_line_info + reset,
-        logging.INFO: cyan + common_format + reset,
-        logging.WARNING: yellow + common_format + reset,
-        logging.ERROR: red + format_with_line_info + reset,
-        logging.CRITICAL: bold_red + format_with_line_info + reset,
-    }
-
-    def format(self, record):
-        """
-        Overriding the base class method to Choose format based on log level.
-        """
-        log_fmt = self.FORMATS.get(record.levelno)
-        formatter = logging.Formatter(log_fmt)
-        return formatter.format(record)
-
-
-def create_logger() -> logging.Logger:
-    """
-    Creates a logger object with Colored QEffFormatter.
-    """
-    logger = logging.getLogger("QEfficient")
-
-    # create console handler and set level to debug
-    ch = logging.StreamHandler()
-    ch.setLevel(logging.INFO)
-    # define formatter
-    ch.setFormatter(QEffFormatter())
-
-    logger.addHandler(ch)
-    return logger
-
-
-# Define the logger object that can be used for logging purposes throughout the module.
-logger = create_logger()
+
+class QEFFLogger:
+    _instance: Optional[logging.Logger] = None
+
+    def __init__(self, loglevel: Optional[str] = "INFO"):
+        if QEFFLogger._instance is None:
+            self.loglevel = loglevel
+            self.logger = self._initialize_logger()
+            QEFFLogger._instance = self.logger
+
+    def _get_formatter(self) -> logging.Formatter:
+        return logging.Formatter(
+            "[%(asctime)s:%(levelname)s:%(threadName)s:%(filename)s:%(lineno)d:%(funcName)s()]:%(message)s"
+        )
+
+    def _initialize_logger(self) -> logging.Logger:
+        # Define the hidden log directory path
+        log_dir = os.path.expanduser("~/.cache/.log")
```

Comment:
should be `QEFF_HOME` set by the user instead of default.
Default would be` ~/.cache/qeff_models`, but can be overridden by the user.